### ISSUE TITLE
SSL routines:ssl_cipher_list_to_bytes:no ciphers available:ssl/statem/statem_clnt.c:3532:No ciphers enabled for max supported SSL/TLS version

### ISSUE THREAD

**Issue by iz8mbw** [openssl/openssl#3931]
Hi.
Using built version (built in /opt/ssldevel) of openssl 1.1.1-tls1.3-draft-18 (source code downloaded with the command: "git clone -b tls1.3-draft-18 --single-branch https://github.com/openssl/openssl.git /path/openssl-1.1.1-tls1.3-draft-18") I was not able to test ciphers by connect to host/port of server. The error seems about CA.

See here:
```
root@server:~# /opt/ssldevel/bin/openssl s_client -cipher ECDHE-RSA-AES256-GCM-SHA384 -connect example.com:443
CONNECTED(00000003)
1995546624:error:141A90B5:SSL routines:ssl_cipher_list_to_bytes:no ciphers available:ssl/statem/statem_clnt.c:3532:No ciphers enabled for max supported SSL/TLS version
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 0 bytes and written 0 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : 0000
    Session-ID:
    Session-ID-ctx:
    Master-Key:
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1500025190
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
---
```

If I execute the same command with OpensSSL 1.1.0f (built in /opt/ssl) it works:

```
root@iz8mbw:~# /opt/ssl/bin/openssl s_client -cipher ECDHE-RSA-AES256-GCM-SHA384 -connect example.com:443
CONNECTED(00000003)
depth=2 C = US, O = DigiCert Inc, OU = www.digicert.com, CN = DigiCert High Assurance EV Root CA
verify return:1
depth=1 C = US, O = DigiCert Inc, OU = www.digicert.com, CN = DigiCert SHA2 High Assurance Server CA
verify return:1
depth=0 C = US, ST = California, L = Los Angeles, O = Internet Corporation for Assigned Names and Numbers, OU = Technology, CN = www.example.org
verify return:1
---
Certificate chain
 0 s:/C=US/ST=California/L=Los Angeles/O=Internet Corporation for Assigned Names and Numbers/OU=Technology/CN=www.example.org
   i:/C=US/O=DigiCert Inc/OU=www.digicert.com/CN=DigiCert SHA2 High Assurance Server CA
 1 s:/C=US/O=DigiCert Inc/OU=www.digicert.com/CN=DigiCert SHA2 High Assurance Server CA
   i:/C=US/O=DigiCert Inc/OU=www.digicert.com/CN=DigiCert High Assurance EV Root CA
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIF8jCCBNqgAwIBAgIQDmTF+8I2reFLFyrrQceMsDANBgkqhkiG9w0BAQsFADBw
MQswCQYDVQQGEwJVUzEVMBMGA1UEChMMRGlnaUNlcnQgSW5jMRkwFwYDVQQLExB3
d3cuZGlnaWNlcnQuY29tMS8wLQYDVQQDEyZEaWdpQ2VydCBTSEEyIEhpZ2ggQXNz
dXJhbmNlIFNlcnZlciBDQTAeFw0xNTExMDMwMDAwMDBaFw0xODExMjgxMjAwMDBa
MIGlMQswCQYDVQQGEwJVUzETMBEGA1UECBMKQ2FsaWZvcm5pYTEUMBIGA1UEBxML
TG9zIEFuZ2VsZXMxPDA6BgNVBAoTM0ludGVybmV0IENvcnBvcmF0aW9uIGZvciBB
c3NpZ25lZCBOYW1lcyBhbmQgTnVtYmVyczETMBEGA1UECxMKVGVjaG5vbG9neTEY
MBYGA1UEAxMPd3d3LmV4YW1wbGUub3JnMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8A
MIIBCgKCAQEAs0CWL2FjPiXBl61lRfvvE0KzLJmG9LWAC3bcBjgsH6NiVVo2dt6u
Xfzi5bTm7F3K7srfUBYkLO78mraM9qizrHoIeyofrV/n+pZZJauQsPjCPxMEJnRo
D8Z4KpWKX0LyDu1SputoI4nlQ/htEhtiQnuoBfNZxF7WxcxGwEsZuS1KcXIkHl5V
RJOreKFHTaXcB1qcZ/QRaBIv0yhxvK1yBTwWddT4cli6GfHcCe3xGMaSL328Fgs3
jYrvG29PueB6VJi/tbbPu6qTfwp/H1brqdjh29U52Bhb0fJkM9DWxCP/Cattcc7a
z8EXnCO+LK8vkhw/kAiJWPKx4RBvgy73nwIDAQABo4ICUDCCAkwwHwYDVR0jBBgw
FoAUUWj/kK8CB3U8zNllZGKiErhZcjswHQYDVR0OBBYEFKZPYB4fLdHn8SOgKpUW
5Oia6m5IMIGBBgNVHREEejB4gg93d3cuZXhhbXBsZS5vcmeCC2V4YW1wbGUuY29t
ggtleGFtcGxlLmVkdYILZXhhbXBsZS5uZXSCC2V4YW1wbGUub3Jngg93d3cuZXhh
bXBsZS5jb22CD3d3dy5leGFtcGxlLmVkdYIPd3d3LmV4YW1wbGUubmV0MA4GA1Ud
DwEB/wQEAwIFoDAdBgNVHSUEFjAUBggrBgEFBQcDAQYIKwYBBQUHAwIwdQYDVR0f
BG4wbDA0oDKgMIYuaHR0cDovL2NybDMuZGlnaWNlcnQuY29tL3NoYTItaGEtc2Vy
dmVyLWc0LmNybDA0oDKgMIYuaHR0cDovL2NybDQuZGlnaWNlcnQuY29tL3NoYTIt
aGEtc2VydmVyLWc0LmNybDBMBgNVHSAERTBDMDcGCWCGSAGG/WwBATAqMCgGCCsG
AQUFBwIBFhxodHRwczovL3d3dy5kaWdpY2VydC5jb20vQ1BTMAgGBmeBDAECAjCB
gwYIKwYBBQUHAQEEdzB1MCQGCCsGAQUFBzABhhhodHRwOi8vb2NzcC5kaWdpY2Vy
dC5jb20wTQYIKwYBBQUHMAKGQWh0dHA6Ly9jYWNlcnRzLmRpZ2ljZXJ0LmNvbS9E
aWdpQ2VydFNIQTJIaWdoQXNzdXJhbmNlU2VydmVyQ0EuY3J0MAwGA1UdEwEB/wQC
MAAwDQYJKoZIhvcNAQELBQADggEBAISomhGn2L0LJn5SJHuyVZ3qMIlRCIdvqe0Q
6ls+C8ctRwRO3UU3x8q8OH+2ahxlQmpzdC5al4XQzJLiLjiJ2Q1p+hub8MFiMmVP
PZjb2tZm2ipWVuMRM+zgpRVM6nVJ9F3vFfUSHOb4/JsEIUvPY+d8/Krc+kPQwLvy
ieqRbcuFjmqfyPmUv1U9QoI4TQikpw7TZU0zYZANP4C/gj4Ry48/znmUaRvy2kvI
l7gRQ21qJTK5suoiYoYNo3J9T+pXPGU7Lydz/HwW+w0DpArtAaukI8aNX4ohFUKS
wDSiIIWIWJiJGbEeIO0TIFwEVWTOnbNl/faPXpk5IRXicapqiII=
-----END CERTIFICATE-----
subject=/C=US/ST=California/L=Los Angeles/O=Internet Corporation for Assigned Names and Numbers/OU=Technology/CN=www.example.org
issuer=/C=US/O=DigiCert Inc/OU=www.digicert.com/CN=DigiCert SHA2 High Assurance Server CA
---
No client certificate CA names sent
Peer signing digest: SHA512
Server Temp Key: ECDH, P-256, 256 bits
---
SSL handshake has read 3388 bytes and written 250 bytes
Verification: OK
---
New, TLSv1.2, Cipher is ECDHE-RSA-AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-GCM-SHA384
    Session-ID: 91F55ABB73C944DCE4DAFB95F45587B38FF3E1A316CCCB0D73F9D3DD15D92779
    Session-ID-ctx:
    Master-Key: 916C18B15E1E633A9BC1F8CFD1A53F1C33CA2647A8CDA26367562394A590C450DE68D52FF7BAB8E6A03D4ED8426C6369
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 300 (seconds)
    TLS session ticket:
    0000 - 36 cf 00 d0 bb 02 6d fe-b0 31 4a f1 e2 df 84 c9   6.....m..1J.....
    0010 - 58 45 5e 00 5f f8 4d 2d-9e f7 b3 22 80 7e 90 69   XE^._.M-...".~.i
    0020 - 04 42 08 3b 9a 7a a9 70-d3 2b b0 56 2b d8 f2 ca   .B.;.z.p.+.V+...
    0030 - c9 13 b2 e1 2e 70 fa 9e-d6 77 da 61 2c 0b d0 c3   .....p...w.a,...
    0040 - d4 69 39 4f a2 67 da 81-4b ea 43 80 81 0b 96 b6   .i9O.g..K.C.....
    0050 - 2f 3c 5f 78 9f af 34 6c-69 1d fe 94 69 ca 12 84   /<_x..4li...i...
    0060 - 01 71 0e 15 c6 cf 4a 43-ba b3 87 05 6c 53 0c d3   .q....JC....lS..
    0070 - b6 d0 91 29 d2 ba 32 27-00 5d 80 06 a8 7a d2 3e   ...)..2'.]...z.>
    0080 - 54 35 dc 74 3d f4 b7 f7-d0 fa ae 45 5d ce 80 bb   T5.t=......E]...
    0090 - 35 45 65 bf 02 93 2c e7-f0 a9 78 22 1d cb e1 0d   5Ee...,...x"....

    Start Time: 1500025391
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
---
```

For your info OpenSSL 1.1.1 tls1.3-draft-18 was configured with: "./config enable-tls1_3 --prefix=/opt/ssldevel --openssldir=/opt/ssldevel -Wl,--enable-new-dtags,-rpath,'$(LIBRPATH)'"

Where is the problem?

Thanks!

**Comment by mattcaswell** [comment]
You get the "No ciphers enabled for max supported SSL/TLS version" error when you have not enabled any ciphersuites that are compatible the highest supported TLS version. So in this case you are restricting the client to only use ECDHE-RSA-AES256-GCM-SHA384 which is not compatible with TLSv1.3. You can either disable TLSv1.3 (using the "-no_tls1_3" s_client option), or additionally provided TLSv1.3 compatible ciphersuites. This issue is discussed in the following blog post (see the Ciphersuites section):

https://www.openssl.org/blog/blog/2017/05/04/tlsv1.3/

**Comment by iz8mbw** [comment]
Ok, thanks, but why by default "openssl s_client" will not automatically use a TLS version different (lower) from 1.3 if that cipher is not available for TLS 1.3 version??

I mean if I go to "openssl s_client" for "ECDHE-RSA-AES256-GCM-SHA384" and this cipher is not available for TLS 1.3 I expect the command will go to use TLS 1.2.
And viceversa, If I go to "openssl s_client" for "TLS13-CHACHA20-POLY1305-SHA256" I expect the command will go to use TLS 1.3.

**Comment by mattcaswell** [comment]
This isn't s_client default behaviour it is libssl default behaviour (which s_client just inherits). We would like to encourage the adoption of TLSv1.3, and we expect it to be a common misconfiguration to accidentally disable all TLSv1.3 ciphersuites when you really did want TLSv1.3. Without this error message you could go on indefinitely with it "just working" without you knowing that you have sub-optimal configuration and never selecting TLSv1.3.
